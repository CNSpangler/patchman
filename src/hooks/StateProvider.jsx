import React, { createContext, useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

const StateProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [body, setBody] = useState('');
  const [method, setMethod] = useState('GET');
  const [results, setResults] = useState('');
  const [history, setHistory] = useState([]);

  const handleUrlChange = ({ target }) => {
    setUrl(target.value);
  };
  
  const handleMethodChange = ({ target }) => {
    setMethod(target.value);
  };
  
  const handleBodyChange = ({ target }) => {
    setBody(target.value);
  };
  
  const handleSubmit = (e) => {
    e.preventDefault();
  
    makeRequest(url, method, body)
      .then(response => setResults(response));
  
    // dependent state change
    setHistory(prevHistory => ([
      ...prevHistory,
      {
        url,
        method,
        body
      }
    ]));
  };

  return (
    <StateContext.Provider value={{ url, body, method, results, history, handleUrlChange, handleMethodChange, handleBodyChange, handleSubmit }}>
      {children}
    </StateContext.Provider>
  );
};

StateProvider.propTypes = {
  children: PropTypes.node
};

export default StateProvider;
