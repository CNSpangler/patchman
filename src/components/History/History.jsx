import React from 'react';
import PropTypes from 'prop-types';
import ReactJson from 'react-json-view';
import styles from '../App/App.css';
import { useHistory } from '../../hooks/StateProvider';

const History = () => {
  const history = useHistory();

  return (
    <div className={styles.History}>
      <h3>History:</h3>
      <ReactJson src={history} />
    </div>
  );
};

History.propTypes = {
  history: PropTypes.arrayOf(PropTypes.shape({
    url: PropTypes.string,
    method: PropTypes.string,
    body: PropTypes.any
  }))
};

export default History;
