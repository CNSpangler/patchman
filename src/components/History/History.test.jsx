import React from 'react';
import { shallow } from 'enzyme';
import History from './History.jsx';
import { StateProvider } from '../../hooks/StateProvider.jsx';

describe('History component', () => {
  it('renders History ', () => {
    const wrapper = shallow(
      <StateProvider>  
        <History />
      </StateProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
