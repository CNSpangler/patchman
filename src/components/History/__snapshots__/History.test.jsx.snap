// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History component renders History  1`] = `
<ContextProvider
  value={
    Object {
      "body": "",
      "handleBodyChange": [Function],
      "handleMethodChange": [Function],
      "handleSubmit": [Function],
      "handleUrlChange": [Function],
      "history": Array [],
      "method": "GET",
      "results": "",
      "url": "",
    }
  }
>
  <History />
</ContextProvider>
`;
